posthog/clickhouse/property_groups.py

```diff
@@ -0,0 +1,72 @@
+from collections.abc import Iterable, MutableMapping
+from dataclasses import dataclass
+
+
+@dataclass
+class PropertyGroupDefinition:
+    filter_expression: str
+    codec: str = "ZSTD(1)"
+
+
+class PropertyGroupManager:
+    def __init__(self, table: str, source_column: str) -> None:
+        self.__table = table
+        self.__source_column = source_column
+        self.__groups: MutableMapping[str, PropertyGroupDefinition] = {}
+
+    def register(self, name: str, definition: PropertyGroupDefinition) -> None:
+        assert name not in self.__groups, "property group names can only be used once"
+        self.__groups[name] = definition
+
+    def __get_map_expression(self, definition: PropertyGroupDefinition) -> str:
+        return f"mapSort(mapFilter((key, _) -> {definition.filter_expression}, CAST(JSONExtractKeysAndValues({self.__source_column}, 'String'), 'Map(String, String)')))"
+
+    def get_alter_table_statements(self, name: str) -> Iterable[str]:
+        column_name = f"{self.__source_column}_group_{name}"
+        definition = self.__groups[name]
+        return [
+            f"ALTER TABLE {self.__table} ADD COLUMN {column_name} Map(String, string) MATERIALIZED {self.__get_map_expression(definition)} CODEC({definition.codec})",
+            f"ALTER TABLE {self.__table} ADD INDEX {column_name}_keys_bf mapValues({column_name}) TYPE bloom_filter",
+            f"ALTER TABLE {self.__table} ADD INDEX {column_name}_values_bf mapValues({column_name}) TYPE bloom_filter",
+        ]
```

Comment:
This will probably need to be tweaked a bit to support `ON CLUSTER` when doing DDL for our sharding setup, this is still a first draft.